### PR TITLE
ValidatingAdmissionPolicy: Factor out matching and params logic for reuse with MutatingAdmissionPolicy

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

This factors out matching and param logic from validatingadmissionpolicy so it can be easily reused for mutatingadmissionpolicy. Taken as the first few leading commits from #123332

#### Special notes for your reviewer:

/assign @cici37

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by cici37** [comment]
The refactor makes sense to me. Thanks!
/lgtm
/hold in case others wanna take a look before merging.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+		newBindingAccessor: newBindingAccessor,
+		matcher:            NewPolicyMatcher(matcher),
+		delegate:           delegate,
+	}
+}
+
+// Dispatch implements generic.Dispatcher. It loops through all active policy-bindings
+// and selects those which are active for the current request. It then resolves
+// all params and creates an Invocation for each matching policy-binding-param tuple.
+// The delegate is then called with the list of tuples.
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:97]:
```suggestion
// The policyDispatcher is then called with the list of tuples.
```

Or am I misreading this?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+) Dispatcher[PolicyHook[P, B, E]] {
+	return &policyDispatcher[P, B, E]{
+		newPolicyAccessor:  newPolicyAccessor,
+		newBindingAccessor: newBindingAccessor,
+		matcher:            NewPolicyMatcher(matcher),
+		delegate:           delegate,
+	}
+}
+
+// Dispatch implements generic.Dispatcher. It loops through all active policy-bindings
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:94]:
```suggestion
// Dispatch implements generic.Dispatcher. It loops through all active policy-bindings hooks
```

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go (hunk 1 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:97]: The policydispatcher organizes all active Hooks (Policy<->Binding) into Invocations (Policy<->Binding<->Params) then calls the delegate with the invocations, so I think the comment is correct

Comment (jpbetz) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/generic/policy_dispatcher.go:97]: I understand now that I've read the implementation.  When I was first looking at the godoc of the exported types I didn't understand what "the delegate" referred to. Is there any way to clarify that?

**Comment by jpbetz** [comment]
/lgtm
Thanks @alexzielenski! Feel free to drop hold when you're ready

**Pull request merged**